### PR TITLE
Adds /creaturetrust command and removes killing of mobs being from container trust.

### PR DESCRIPTION
**Pull request by Skizzles**
Adds /creaturetrust command, along with messages, database storage for it, and adds it to all other areas. This separates killing of mobs from the containertrust permission.

This is compatible with old installations and has been tested as working, if players use the new command the information is automatically added and updated accordingly with their claim.

Feel free to let me know if I am missing something or need to change something!

### CODE REVIEW

**Comment by RoboMWM** [comment]
I appreciate the PR, however, I prefer to keep the trust system as simple as possible. I know I've been slow on refactoring for a better API, but extra trust levels is something I also want to give out to addons, so addons can implement permissions in any way they choose (be it extra trust levels, or even more granular with flags, etc.)

**Comment by Skizzles** [comment]
> I appreciate the PR, however, I prefer to keep the trust system as simple as possible. I know I've been slow on refactoring for a better API, but extra trust levels is something I also want to give out to addons, so addons can implement permissions in any way they choose (be it extra trust levels, or even more granular with flags, etc.)

While I understand keeping things simple, the current permissions being how they are open players up to grief by forcing them to allow container trust, making it required to have a secondary plugin to protect containers. I created this so players could also have public areas for grinders which was not possible with the current setup. All other config options, commands, etc opened players up to grief or having other animals killed.

If examples of necessity or more opinions would get this PR merged I can gather that, but I truly believe this would be a great addition in the meantime before an addon system and API in a stable place.

File: src/main/java/me/ryanhamshire/GriefPrevention/FlatFileDataStore.java [hunk 1]
```diff
@@ -320,6 +320,11 @@ else if (this.getSchemaVersion() == 0)
                         List<String> containerNames = Arrays.asList(line.split(";"));
                         containerNames = this.convertNameListToUUIDList(containerNames);
 
+                        //fifth line is list of players who can access creatures
+                        line = inStream.readLine();
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/FlatFileDataStore.java:324]: This is the sixth line (not fifth, as the copied comment indicates), which is already used for access trust data. This unexpectedly elevates permissions for users.

File: src/main/java/me/ryanhamshire/GriefPrevention/FlatFileDataStore.java [hunk 2]
```diff
@@ -320,6 +320,11 @@ else if (this.getSchemaVersion() == 0)
...
+                        List<String> creatureNames = Arrays.asList(line.split(";"));
+                        creatureNames = this.convertNameListToUUIDList(creatureNames);
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/FlatFileDataStore.java:326]: Name list should already be a UUID list - this conversion is for legacy data. Doesn't need to be done.

File: src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java [hunk 3]
```diff
@@ -101,7 +101,7 @@ void initialize() throws Exception
         {
             //ensure the data tables exist
             statement.execute("CREATE TABLE IF NOT EXISTS griefprevention_nextclaimid (nextid INTEGER)");
-            statement.execute("CREATE TABLE IF NOT EXISTS griefprevention_claimdata (id INTEGER, owner VARCHAR(50), lessercorner VARCHAR(100), greatercorner VARCHAR(100), builders TEXT, containers TEXT, accessors TEXT, managers TEXT, inheritnothing BOOLEAN, parentid INTEGER)");
+            statement.execute("CREATE TABLE IF NOT EXISTS griefprevention_claimdata (id INTEGER, owner VARCHAR(50), lessercorner VARCHAR(100), greatercorner VARCHAR(100), builders TEXT, containers TEXT, creatures, TEXT, accessors TEXT, managers TEXT, inheritnothing BOOLEAN, parentid INTEGER)");
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java]: `creatures, TEXT` -> `creatures TEXT`

File: src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java [hunk 4]
```diff
@@ -47,7 +47,7 @@ public class DatabaseDataStore extends DataStore
     private static final String SQL_UPDATE_NAME =
             "UPDATE griefprevention_playerdata SET name = ? WHERE name = ?";
     private static final String SQL_INSERT_CLAIM =
-            "INSERT INTO griefprevention_claimdata (id, owner, lessercorner, greatercorner, builders, containers, accessors, managers, inheritnothing, parentid) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
+            "INSERT INTO griefprevention_claimdata (id, owner, lessercorner, greatercorner, builders, containers, creatures, accessors, managers, inheritnothing, parentid) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java]: Missing corresponding new VALUES entry.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java [hunk 5]
```diff
@@ -214,7 +214,7 @@ boolean hasSurfaceFluids()
     }
 
     //main constructor.  note that only creating a claim instance does nothing - a claim must be added to the data store to be effective
-    Claim(Location lesserBoundaryCorner, Location greaterBoundaryCorner, UUID ownerID, List<String> builderIDs, List<String> containerIDs, List<String> accessorIDs, List<String> managerIDs, boolean inheritNothing, Long id)
+    Claim(Location lesserBoundaryCorner, Location greaterBoundaryCorner, UUID ownerID, List<String> builderIDs, List<String> containerIDs, List<String> creatureIDs, List<String> accessorIDs, List<String> managerIDs, boolean inheritNothing, Long id)
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:217]: Needs an overload to not break existing addons using this constructor.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java [hunk 6]
```diff
@@ -256,9 +261,9 @@ boolean hasSurfaceFluids()
         this.inheritNothing = inheritNothing;
     }
 
-    Claim(Location lesserBoundaryCorner, Location greaterBoundaryCorner, UUID ownerID, List<String> builderIDs, List<String> containerIDs, List<String> accessorIDs, List<String> managerIDs, Long id)
+    Claim(Location lesserBoundaryCorner, Location greaterBoundaryCorner, UUID ownerID, List<String> builderIDs, List<String> containerIDs, List<String> creatureIDs, List<String> accessorIDs, List<String> managerIDs, Long id)
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:264]: Needs an overload.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java [hunk 7]
```diff
@@ -710,7 +715,7 @@ public void clearPermissions()
 
     //gets ALL permissions
     //useful for  making copies of permissions during a claim resize and listing all permissions in a claim
-    public void getPermissions(ArrayList<String> builders, ArrayList<String> containers, ArrayList<String> accessors, ArrayList<String> managers)
+    public void getPermissions(ArrayList<String> builders, ArrayList<String> containers, ArrayList<String> creatures, ArrayList<String> accessors, ArrayList<String> managers)
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:718]: Needs an overload.

File: src/main/java/me/ryanhamshire/GriefPrevention/ClaimPermission.java [hunk 8]
```diff
@@ -38,6 +38,10 @@
     /**
      * ClaimPermission used for basic access.
      */
+    Creatures(Messages.NoCreaturesPermission),
+    /**
+     * ClaimPermission used for attacking mobs.
+     */
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/ClaimPermission.java:44]: Comment is beneath the enum value, which was also inserted beneath the access trust comment.

File: src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java [hunk 9]
```diff
@@ -339,14 +340,18 @@ else if (this.getSchemaVersion() < 1)
                 List<String> containerNames = Arrays.asList(containersString.split(";"));
                 containerNames = this.convertNameListToUUIDList(containerNames);
 
+                String creaturesString = results.getString("creatures");
+                List<String> creaturesNames = Arrays.asList(creaturesString.split(";"));
+                creaturesNames = this.convertNameListToUUIDList(creaturesNames);
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java]: Name list should already be a UUID list - this conversion is for legacy data. Doesn't need to be done.

**Review by Jikoo — CHANGES_REQUESTED** [review]
This is not compatible with existing installations, potentially incompatible with addons depending on which methods they use, and the database datastore is broken. See review comments for specifics.

As far as personal reservations go, this adds confusing complexity. Most of the requests I've seen for creature trust are for allowing only access to animals, i.e. milking cows and shearing sheep and not killing, so I believe most users would expect that to be what this is. GP's linear trust model makes it difficult for specific behavioral changes like that to be conveyed to users, especially once addons get involved. I believe in GP's current state this is best entirely left to an addon rather than partially because it will reduce player confusion from server to server.
> the current permissions being how they are open players up to grief by forcing them to allow container trust, making it required to have a secondary plugin to protect containers. I created this so players could also have public areas for grinders which was not possible with the current setup.

Create a subclaim around the pen/kill zone and only containertrust people inside it. They won't be able to access the storage chests in the rest of the claim, but they will be able to kill things. Bonus points if you allow them to use a lever to disable hoppers in the grinder so they can pick up loot instead of having it go into a sorting system.

File: src/main/java/me/ryanhamshire/GriefPrevention/FlatFileDataStore.java [hunk 10]
```diff
@@ -320,12 +320,17 @@ else if (this.getSchemaVersion() == 0)
                         List<String> containerNames = Arrays.asList(line.split(";"));
                         containerNames = this.convertNameListToUUIDList(containerNames);
 
-                        //sixth line is list of players who can use buttons and switches
+                        //sixth line is list of players who can access creatures
+                        line = inStream.readLine();
```
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/FlatFileDataStore.java:324]: You didn't actually resolve the problems here by correcting the line number comments. This elevates existing access trust and converts existing permission trust into access trust. A new type of trust should be the last line. You cannot insert it into the middle without providing a converter for existing data.

**Review by Jikoo — CHANGES_REQUESTED** [review]

**Comment by RoboMWM** [comment]
I understand there is a need for more granular permission tiers on some servers - and as such, I'd be more preferential to PRs that help refactor/add API capabilities to this if it is not already possible.

**Pull request closed without merging**